sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala

```diff
@@ -62,7 +62,30 @@ private[sql] object DataSourceStrategy extends Strategy with Logging {
     // Scanning partitioned HadoopFsRelation
     case PhysicalOperation(projects, filters, l @ LogicalRelation(t: HadoopFsRelation))
         if t.partitionSpec.partitionColumns.nonEmpty =>
-      val selectedPartitions = prunePartitions(filters, t.partitionSpec).toArray
+      // We divide the filter expressions into 3 parts
+      val partitionColumnNames = t.partitionSpec.partitionColumns.map(_.name).toSet
+      val filterMap = filters.groupBy { f =>
+        // TODO this is case-senstive
+        val referencedColumnNames = f.references.map(_.name).toSet
+        if (referencedColumnNames.subsetOf(partitionColumnNames)) {
+          // Only reference the partition key
+          0
+        } else if (referencedColumnNames.intersect(partitionColumnNames).isEmpty) {
+          // Not reference any partition key at all. can be push down
+          1
+        } else {
+          // Reference both partition key and attributes
+          2
+        }
+      }
+      // Only prunning the partition keys
+      val partitionFilters = filterMap.getOrElse(0, Nil)
+      // Only pushes down predicates that do not reference partition keys.
+      val pushedFilters = filterMap.getOrElse(1, Nil)
+      // Predicates with both partition keys and attributes
+      val combineFilters = filterMap.getOrElse(2, Nil)
```

Comment:
Instead of using `groupBy`, can we just use `filter` 3 times to split these 3 kinds of filters? I think performance doesn't matter here.